src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -1147,6 +1193,707 @@ bool Compiler::optMakeExitTestDownwardsCounted(ScalarEvolutionContext& scevConte
+
+        m_cursors1.Emplace(block, stmt, tree, primaryIV, isInsideExitTest);
+        m_cursors2.Emplace(block, stmt, tree, primaryIV, isInsideExitTest);
+        return true;
+    };
+
+    if (!m_loopLocals.VisitOccurrences(m_loop, primaryIVLcl->GetLclNum(), visitor) || (m_cursors1.Height() <= 0))
+    {
+        JITDUMP("  Could not create cursors for all loop uses of primary IV");
+        return false;
+    }
+
+    JITDUMP("  Found %d cursors using primary IV V%02u\n", m_cursors1.Height(), primaryIVLcl->GetLclNum());
+
+#ifdef DEBUG
+    if (m_comp->verbose)
+    {
+        for (int i = 0; i < m_cursors1.Height(); i++)
+        {
+            CursorInfo& cursor = m_cursors1.BottomRef(i);
+            printf("    [%d] [%06u]%s: ", i, Compiler::dspTreeID(cursor.Tree),
+                   cursor.IsInsideExitTest ? " (in-test)" : "");
+            cursor.IV->Dump(m_comp);
+            printf("\n");
+        }
+    }
+#endif
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// AdvanceCursors: Advance cursors stored in "cursors" and store the advanced
+// result in "nextCursors".
+//
+// Parameters:
+//   cursors     - [in] List of current cursors. Unmodified.
+//   nextCursors - [in, out] List of next cursors. The "Tree" and "IV" fields
+//                 of these cursors will be updated to point to the next derived
+//                 IV.
+//
+// Returns:
+//   True if all uses were analyzed and cursors could be introduced for them
+//   all; otherwise false.
+//
+// Remarks:
+//   A cursor is created for a use when it represents the same value as the
+//   primary IV passed. The function will allow mismatching uses if the use is
+//   expected to be removed in the downwards loop transformation. Otherwise the
+//   function will fail.
+//
+//   It is not a correctness requirement that we remove all uses; if we end up
+//   not doing so (e.g. because a cursor was not created by this function),
+//   then we may just end up with extra primary IVs in the loop.
+//
+void StrengthReductionContext::AdvanceCursors(ArrayStack<CursorInfo>* cursors, ArrayStack<CursorInfo>* nextCursors)
+{
+    for (int i = 0; i < cursors->Height(); i++)
+    {
+        CursorInfo& cursor     = cursors->BottomRef(i);
+        CursorInfo& nextCursor = nextCursors->BottomRef(i);
+
+        assert((nextCursor.Block == cursor.Block) && (nextCursor.Stmt == cursor.Stmt) &&
+               (nextCursor.IsInsideExitTest == cursor.IsInsideExitTest));
+
+        nextCursor.Tree = cursor.Tree;
+        do
+        {
+            GenTree* cur    = nextCursor.Tree;
+            nextCursor.Tree = cur->gtGetParent(nullptr);
+
+            if ((nextCursor.Tree == nullptr) ||
+                (nextCursor.Tree->OperIs(GT_COMMA) && (nextCursor.Tree->gtGetOp1() == cur)))
+            {
+                nextCursor.IV = nullptr;
+                break;
+            }
+
+            // TODO-CQ: If this is now the source to a store, we can
```

Comment:
Do you know how often you see this? I suppose for normal array uses it's rare, we may only see it for pointer math where users have CSE'd parts of the computation or something.

Reply:
Fairly often:
```
Advanced cursor node type
ADD                  :   26343
CAST                 :   15729
LSH                  :   11239
BOUNDS_CHECK         :    8355
IND                  :    5714
LT                   :    3143
GT                   :    2536
STORE_LCL_VAR        :    1319
GE                   :    1252
SUB                  :    1008
CALL                 :     993
STORE_BLK            :     907
HWINTRINSIC          :     892
BLK                  :     841
UMOD                 :     822
LE                   :     437
STOREIND             :     357
EQ                   :     297
MUL                  :      89
AND                  :      63
NE                   :      61
MOD                  :      38
INTRINSIC            :      16
NEG                  :      14
UDIV                 :      10
COMMA                :       7
BITCAST              :       2
DIV                  :       2
RSH                  :       2
```